### PR TITLE
Help command fix, normalize suggestions for unknown commands

### PR DESCRIPTION
**Pull request by RohanJnr**
## Relevant Issues
issue: #863 

Refer to https://github.com/python-discord/sir-lancebot/pull/884

## Fixes
- The help command used to raise an error when provided with an invalid command name (fixed)
- There used to be different command suggestions from the help command and the error handler when an unknown command was provided.
  - This has been fixed by using the same method for finding similar commands.

## Changes
- Providing an invalid subcommand but valid parent command (for example: `!ext listt`, ext is valid but subcommand `listt` is invalid) now sends the help command output for the parent command.

- Using `bot.all_commands` attribute instead of `bot.walk_commands()` generator.
  - `walk_commands()` is not useful in the error_handler because (as it provides us with all subcommands also), when there is an invalid subcommand, the parent command (`commands.Group`) handles this case and sends a help message, hence, never reaching the error_handler.
  - `all_commands` does not give us subcommands (it gives aliases which is pretty useful).

- Using `rapidfuzz` for finding similar commands instead of `difflib` (reasons being speed).

- Error handler now sends multiple command suggestions instead of 1 when user is trying to run an unknown command.

## Yet to decide/to discuss
The help command lists/suggests commands that may not in the user's power to invoke them. Should we filter commands when listing them such that to only show them the commands which they can invoke ?

This is the current output when user invokes help command on a restricted command.

## Did you:

- [X] Join the [**Python Discord Community**](https://discord.gg/python)?
- [X] Read all the comments in this template?
- [X] Ensure there is an issue open, or link relevant discord discussions?
- [X] Read and agree to the [contributing guidelines](https://pythondiscord.com/pages/contributing/contributing-guidelines/)?

### CODE REVIEW

File: bot/exts/core/error_handler.py [hunk 1]
```diff
@@ -158,31 +158,32 @@ async def on_command_error(self, ctx: commands.Context, error: commands.CommandE
...
-            log_msg = "Cancelling attempt to suggest a command due to failed checks."
-            try:
-                if not await similar_command.can_run(ctx):
+        command_suggestions = []
+        if similar_command_names := get_command_suggestions(list(self.bot.all_commands.keys()), command_name):
+            for similar_command_name in similar_command_names:
+                similar_command = self.bot.get_command(similar_command_name)
+
+                if not similar_command:
+                    return
```
Comment (Bluenix2) [inline bot/exts/core/error_handler.py:167]: There are other similar command names we can try if we can't find this one, I would suggest changing this to `continue`:
```suggestion
                    continue
```

File: bot/exts/core/error_handler.py [hunk 2]
```diff
@@ -158,31 +158,32 @@ async def on_command_error(self, ctx: commands.Context, error: commands.CommandE
...
+
+                log_msg = "Cancelling attempt to suggest a command due to failed checks."
+                try:
+                    if not await similar_command.can_run(ctx):
+                        log.debug(log_msg)
+                        return
```
Comment (Bluenix2) [inline bot/exts/core/error_handler.py:173]: Same as above, we can assume that there may be other suggestions:
```suggestion
                        continue
```

File: bot/exts/core/help.py [hunk 3]
```diff
@@ -115,6 +119,9 @@ def __init__(
     def _get_query(self, query: str) -> Union[Command, Cog]:
         """Attempts to match the provided query with a valid command or cog."""
         command = self._bot.get_command(query)
+
+        log.info(query)
+        log.info(command)
```
Comment (Bluenix2) [inline bot/exts/core/help.py:124]: Were these just local debugs? They don't make a lot of sense without any explanation. I think it is best if we just remove them?

File: bot/exts/core/help.py [hunk 4]
```diff
@@ -153,12 +160,17 @@ def _handle_not_found(self, query: str) -> None:
 
         Will pass on possible close matches along with the `HelpQueryNotFound` exception.
         """
-        # Combine command and cog names
-        choices = list(self._bot.all_commands) + list(self._bot.cogs)
+        # Check if parent command in valid incase subcommand is invalid.
+        if " " in query:
+            parent, sub_command = query.split()
```
Comment (Bluenix2) [inline bot/exts/core/help.py:165]: Don't we have deeper nested subcommands than once? I think you should pass `maxsplit=1`.

**Review by brad90four — APPROVED** [review]
Content wise this looks good to me, have not verified with independent testing yet.

File: bot/exts/core/help.py [hunk 5]
```diff
@@ -41,14 +41,18 @@ class HelpQueryNotFound(ValueError):
...
-    query, where keys are the possible matched command names and values are the likeness match scores.
+    Params:
+    possible_matches: list of similar command names.
+    parent_command: parent command of an invalid subcommand. Only available when an invalid subcommand
+                    has been passed.
     """
 
-    def __init__(self, arg: str, possible_matches: dict = None):
+    def __init__(
+            self, arg: str, possible_matches: list[str] = None, *, parent_command: Command = None
```
Comment (GDWR) [inline bot/exts/core/help.py:51]: Missing `Optional` or `| None` typing.
```suggestion
            self, arg: str, possible_matches: list[str] | None = None, *, parent_command: Command | None = None
```

**Review by GDWR — CHANGES_REQUESTED** [review]
Looks good, just a slight type hint that could catch somebody out.

**Comment by ichard26** [comment]
I'm in the middle of a review (sorry for taking so long!) but I ran out of time to finish it today :/ I'll submit my review tomorrow.

File: bot/exts/core/help.py (hunk 5 above)
Comment (RohanJnr) [reply bot/exts/core/help.py:51]: Good catch, though we need to use `Optional` from the `typing` module since the project is running on 3.9 and not 3.10.

**Review by GDWR — COMMENTED** [review]
Looks good, I'll leave the final approval to @ichard26 so it doesn't auto-merge. 
(I think that is how this works)

**Review by ichard26 — CHANGES_REQUESTED** [review]
I wasn't able to find time to do a full review with manual testing (and I feel bad about delaying the PR even more), so consider this just a read-through review. On that note, it looks good except for a poorly written comment.

**Review by wookie184 — APPROVED** [review]
Had a quick look over, looks good and seems to work nicely, thanks 👍

**Pull request merged**